Enable toggling comments on/off in Expression Builder

JuhoErvasti opened the pull request:
Fixes #61356.

Implements `toggleComment` for `QgsCodeEditorExpression` and adds a new button to the Expression Builder widget to toggle comments for selected lines.

![exprtest](https://github.com/user-attachments/assets/5d87fdd8-9eaa-4d6e-a6c7-ffd92a1d97e4)

JuhoErvasti committed:
Implement toggleComment for QgsCodeEditorExpression

JuhoErvasti committed:
Add test case for QgsCodeEditorExpression::toggleComment

JuhoErvasti committed:
Add button to toggle comments in expression builder widget

nyalldawson commented on src/gui/codeeditors/qgscodeeditorexpression.h:
```diff
@@ -56,6 +56,11 @@ class GUI_EXPORT QgsCodeEditorExpression : public QgsCodeEditor
      */
     void setFields( const QgsFields &fields );
 
+    /**
+     * Toggles comments for selected lines.
+     */
```
We don't do documentation for overridden methods:

```suggestion
```

nyalldawson commented on src/gui/codeeditors/qgscodeeditorexpression.h:
```diff
@@ -56,6 +56,11 @@ class GUI_EXPORT QgsCodeEditorExpression : public QgsCodeEditor
+    /**
+     * Toggles comments for selected lines.
+     */
+    void toggleComment() override;
```
As per the docs for QgsCodeEditor::toggleComment, you'll also need to override QgsCodeEditor::languageCapabilities() and return the ToggleComment capability.

nyalldawson commented on src/gui/codeeditors/qgscodeeditorexpression.cpp:
```diff
@@ -36,6 +36,87 @@ Qgis::ScriptLanguage QgsCodeEditorExpression::language() const
 
+void QgsCodeEditorExpression::toggleComment()
+{
+  if ( isReadOnly() )
```
This looks like 99% copy and paste from QgsCodeEditorPython::toggleComment. How about moving this to a protected method `QgsCodeEditor::toggleLineComments( const QString& commentPrefix )` instead?

nyalldawson commented:
How do you feel about moving the button up to the toolbar? That'd give consistency with the other code editors.

![image](https://github.com/user-attachments/assets/4d885058-bba1-47bc-a7bb-72d56514a9fc)

JuhoErvasti committed:
Update src/gui/codeeditors/qgscodeeditorexpression.h

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

JuhoErvasti committed:
Address review

JuhoErvasti commented:
@nyalldawson Thank you for the review. I moved the button upwards.

I also noticed a bug with the comment toggling which I think has existed in the python code editor for a while, where if your selection starts or ends on an empty line the selection shifts incorrectly, I've fixed that too. Couldn't find an issue for that.

JuhoErvasti committed:
Remove unnecessary check

nyalldawson commented:
Thanks @JuhoErvasti !

DelazJ commented:
> How do you feel about moving the button up to the toolbar? That'd give consistency with the other code editors.

@nyalldawson what other code editors did you have in mind? Do they also have top and bottom toolbars?
That location looks weird to me in this dialog:
* most of the tools in the top toolbar are for saving/managing user expressions, except the "Clear expression editor" button. And if the "--" button were to be put in the top toolbar, I'd rather see it close to the "clear" button, not after a set of unrelated but specific buttons.
* the ones at the bottom are tools that actually help to write/format the code, what `--` button also is.

Opinions?

nyalldawson commented:
@DelazJ 

> what other code editors did you have in mind? 

eg  processing script editor:

<img width="823" height="140" alt="image" src="https://github.com/user-attachments/assets/1c4818d8-2df1-4a5e-b1cc-3a21b007b478" />

python console:

<img width="957" height="47" alt="image" src="https://github.com/user-attachments/assets/3c67104c-0002-416e-afe2-420735922655" />

> most of the tools in the top toolbar are for saving/managing user expressions, except the "Clear expression editor" button. And if the "--" button were to be put in the top toolbar, I'd rather see it close to the "clear" button, not after a set of unrelated but specific buttons.

Ideally we'd rework that whole top bar -- it should show all the standard buttons for code editors, like cut/copy/paste/undo/redo/toggle find. (Just like we have in the processing script editor from the screenshot above, or in the Execute SQL dialog below):

<img width="510" height="85" alt="image" src="https://github.com/user-attachments/assets/5ffe66ac-c45e-4905-85cf-addd17a2b924" />

> the ones at the bottom are tools that actually help to write/format the code, what -- button also is.

Well - I'd argue that they are all about INSERTING bits into the expression -- not modifying what you already have. If we exposed a "find and replace" function in this dialog then it would definitely belong in the top toolbar, and that's similar in behavior to toggling comments (it changes existing content, not inserting preset strings).
